### PR TITLE
PyTorch DistributedDataParallel Multi-GPU training

### LINKED ISSUE #1332: PyTorch multi-GPU training, multi-node or single node, async or synced
For the PyTorch engine (#1120).

Multi-GPU training. Multi-node or single node. Async vs sync. It should also work well with consumer GPUs, also in the multi-node case. Currently distributed data parallel synchronized training, using `DistributedDataParallel`, and using our own dataset sampling logic, extending `Dataset._get_default_random_seed_offset`.

This issue is to track the process, discuss anything relevant.

### PR DESCRIPTION
**Pull request by Judyxujj**
For PyTorch (#1120).

1. Multi GPU training with `DistributedDataParallel`
2. Gradient accumulation
3. Add elapsed training time for each sub-epoch

Fix #1332.

Question: how do we set the multi-GPU flag in the returnn config?  I saw for the automated mix precision, the opts dict should be defined in the returnn config. Should I also use a dict opts for the DistributedDataParallel in the returnn config? So that we can define parameters like `bucket_cap_mb`, `find_unused_parameters` in the opts dict?

### CODE REVIEW

File: returnn/torch/engine.py [hunk 1]
```diff
@@ -64,6 +68,10 @@ def __init__(self, config: Config):
         self._device = _get_device_from_config(config)
         print("Using device:", self._device, file=log.v2)
 
+        self._use_DDP = config.is_true("use_DDP")  # type: Optional[bool]
```
Comment (albertz) [inline returnn/torch/engine.py:71]: As discussed, we want to have this a dict, and rename it, like `torch_distributed`.

Comment (albertz) [reply returnn/torch/engine.py:71]: There would be one `class` or `type` inside that dict, which would be `"DistributedDataParallel"`.

File: returnn/torch/engine.py
Comment (albertz) [inline returnn/torch/engine.py:1]: I think we should move out the DDP code into a separate file as much as possible, similar to how we have `returnn/tf/horovod.py` as a separate file, and then similar as `HorovodContext`, have some `DistributedContext` or so.

File: returnn/torch/engine.py [hunk 2]
```diff
@@ -105,6 +113,11 @@ def init_train_from_config(
         """
         assert config is self.config
         super().init_train_from_config(config=config)
+        if self._use_DDP:
+            ddp_local_rank = int(os.environ["LOCAL_RANK"])
```
Comment (albertz) [inline returnn/torch/engine.py:117]: You should not use the `os.environ` but rather the `torch.distributed` functions.

File: returnn/torch/engine.py [hunk 3]
```diff
@@ -130,6 +143,13 @@ def init_train_from_config(
         self._load_model(epoch=self._start_epoch)
         self._save_model_epoch_interval = config.int("save_interval", 1)
 
+        if self._use_DDP:
+            # wrap the model use DDP
+            self._pt_model = DDP(
+                self._pt_model,
+                device_ids=[int(os.environ["LOCAL_RANK"])],
+                find_unused_parameters=self._DDP_find_unused_parameters,
```
Comment (albertz) [inline returnn/torch/engine.py:151]: Instead of passing this, you would pass all the options from the user.

File: returnn/torch/engine.py [hunk 4]
```diff
@@ -138,6 +158,11 @@ def init_train_from_config(
         self._train_step_func = self.config.typed_value("train_step")
         assert self._train_step_func, "train_step not defined"
 
+        self._gradient_accumulation_steps = config.int("gradient_accumulation_steps", 1)
```
Comment (albertz) [inline returnn/torch/engine.py:161]: In the TF engine, the option is called `accum_grad_multiple_step`. Why do you call it different here? I think we should just use the same name.

File: returnn/torch/engine.py [hunk 5]
```diff
@@ -138,6 +158,11 @@ def init_train_from_config(
...
+        if self._gradient_accumulation_steps > 1:
+            assert self._gradient_accumulation_steps % torch.cuda.device_count() == 0
+            self._gradient_accumulation_steps //= torch.cuda.device_count()
```
Comment (albertz) [inline returnn/torch/engine.py:164]: As discussed, I think this is unexpected, so we should not do this.

File: returnn/torch/engine.py [hunk 6]
```diff
@@ -359,9 +424,70 @@ def _run_step(self, extern_data_raw: Dict[str, torch.Tensor], *, train_flag: boo
...
+                    # this forward pass, to ensure we short circuit reduction for any
+                    # unused parameters. Only if `find_unused_parameters` is set.
+                    if self._pt_model.find_unused_parameters and not self._pt_model.static_graph:
+                        # Do not need to populate this for static graph.
+                        self._pt_model.reducer.prepare_for_backward(
+                            list(torch.nn.parallel.distributed._find_tensors(output))
+                        )
+                    else:
+                        self._pt_model.reducer.prepare_for_backward([])
+                else:
```
Comment (albertz) [inline returnn/torch/engine.py:484]: I assume you just copied all this code? Did you somehow adapt this?

You are calling a few private functions here. Not sure if this is a good idea.

In any case, we should move this code out here, into some context manager helper.

File: returnn/torch/engine.py [hunk 7]
```diff
@@ -359,9 +424,70 @@ def _run_step(self, extern_data_raw: Dict[str, torch.Tensor], *, train_flag: boo
...
+
+                if self._pt_model._join_config.enable:
+                    # Notify joined ranks whether they should sync in backwards pass or not.
+                    self._pt_model._check_global_requires_backward_grad_sync(is_joined_rank=False)
+                with self._pt_model._inside_ddp_forward():
+                    with autocast(
+                        device_type=self._device, dtype=self._autocast_dtype
+                    ) if self._use_autocast else nullcontext():
+                        # If the flag use_DDP is enabled, we need to return the forward output in the train step function
+                        output = self._train_step_func(model=self._orig_model, extern_data=extern_data, **sentinel_kw)
```
Comment (albertz) [inline returnn/torch/engine.py:465]: There is no output.

File: returnn/torch/engine.py [hunk 8]
```diff
@@ -359,9 +424,70 @@ def _run_step(self, extern_data_raw: Dict[str, torch.Tensor], *, train_flag: boo
...
+                    self._pt_model.require_forward_param_sync = True
+                    # We'll return the output object verbatim since it is a freeform
+                    # object. We need to find any tensors in this object, though,
+                    # because we need to figure out which parameters were used during
+                    # this forward pass, to ensure we short circuit reduction for any
+                    # unused parameters. Only if `find_unused_parameters` is set.
+                    if self._pt_model.find_unused_parameters and not self._pt_model.static_graph:
+                        # Do not need to populate this for static graph.
+                        self._pt_model.reducer.prepare_for_backward(
+                            list(torch.nn.parallel.distributed._find_tensors(output))
```
Comment (albertz) [inline returnn/torch/engine.py:480]: Use the loss instead. Access it via `rf.get_run_ctx()...`.

File: returnn/datasets/basic.py [hunk 9]
```diff
@@ -245,6 +245,11 @@ def _get_default_random_seed_offset():
 
             if returnn.tf.horovod.get_ctx().is_dataset_distribution_random_seed_offset():
                 return returnn.tf.horovod.get_ctx().rank() * 16127
+
+        if config.is_true("use_DDP"):
```
Comment (albertz) [inline returnn/datasets/basic.py:249]: `torch_distributed`

File: returnn/log.py [hunk 10]
```diff
@@ -192,6 +192,15 @@ def init_by_config(self, config):
                 fn_ext = ".horovod-%i-%i%s" % (hvd.rank(), hvd.size(), fn_ext)
                 new_logs.append(fn_prefix + fn_ext)
             logs = new_logs
+
+        if config.is_true("use_DDP"):
+            new_logs = []
+            for fn in logs:
+                fn_prefix, fn_ext = os.path.splitext(fn)
+                fn_ext = ".nccl-%i-%i%s" % (int(os.environ["LOCAL_RANK"]), int(os.environ["WORLD_SIZE"]), fn_ext)
```
Comment (albertz) [inline returnn/log.py:200]: Why "nccl"? Call it "torch-distrib" or so. Also, do use `os.environ`.

File: returnn/torch/engine.py (hunk 6 above)
Comment (Judyxujj) [reply returnn/torch/engine.py]: Yes, I have copied the codes. Where should be put the context manager helper, I now put it under returnn.torch.distributed

File: returnn/log.py (hunk 10 above)
Comment (Judyxujj) [reply returnn/log.py]: I have to use `os.environ` to get the local rank, the `get_rank()` function from torch.distributed return rank and I couldn't find other function from torch.distributed to get the local rank

File: returnn/torch/engine.py [hunk 11]
```diff
@@ -400,6 +526,8 @@ def _load_model(self, *, epoch: int):
         print("Model:", self._pt_model, file=log.v4)
 
         if checkpoint_state is not None:
+            # remove the prefix of "module." from the DDP state dict
+            torch.nn.modules.utils.consume_prefix_in_state_dict_if_present(checkpoint_state["model"], "module.")
```
[the remark this answers is not in the document]
Comment (Judyxujj) [reply returnn/torch/engine.py:498]: I am not sure in which cases we would need to keep the prefix module, if there are such cases, then I can add a condition here

File: returnn/torch/engine.py (hunk 8 above)
Comment (albertz) [reply returnn/torch/engine.py]: I don't really understand what you mean. What do you mean by "the variable would be marked twice"? What does it mean to mark a variable?

To me it sounds like there is still some issue in the way we wrap the `forward` call.

File: returnn/__main__.py [hunk 12]
```diff
@@ -499,12 +508,12 @@ def execute_main_task():
         engine.init_network_from_config(config)
         engine.web_server(port=config.int("web_server_port", 12380))
     elif task.startswith("config:"):
-        action = config.typed_dict[task[len("config:") :]]
+        action = config.typed_dict[task[len("config:"):]]
```
Comment (albertz) [inline returnn/__main__.py:511]: Why this change?

**Comment by albertz** [comment]
Where is the `returnn/torch/distributed.py` file?

File: returnn/torch/engine.py [hunk 13]
```diff
@@ -155,6 +185,8 @@ def train(self):
             self.epoch += 1
             self._epoch_mp_shared.value = self.epoch
 
+        if self._use_torch_distributed:
+            destroy_process_group()
```
Comment (albertz) [inline returnn/torch/engine.py:189]: This should be moved to `torch/distributed.py` as well.

File: returnn/torch/engine.py [hunk 14]
```diff
@@ -359,9 +438,24 @@ def _run_step(self, extern_data_raw: Dict[str, torch.Tensor], *, train_flag: boo
...
 
-        with autocast(device_type=self._device, dtype=self._autocast_dtype) if self._use_autocast else nullcontext():
-            sentinel_kw = {"__fwd_compatible_random_arg_%i" % int(random() * 100): None}
-            self._train_step_func(model=self._orig_model, extern_data=extern_data, **sentinel_kw)
+        sentinel_kw = {"__fwd_compatible_random_arg_%i" % int(random() * 100): None}
+
+        with autocast(
+            device_type=self._device, dtype=self._autocast_dtype
+        ) if self._use_autocast else nullcontext():
+            if not self._use_torch_distributed:
```
Comment (albertz) [inline returnn/torch/engine.py:446]: There should be no if-statement here. You always should call the same `_train_step_func` func, in the same way.

The if-statement should be for the context manager, so inside the `with` statement above.

File: returnn/torch/engine.py [hunk 15]
```diff
@@ -359,9 +438,24 @@ def _run_step(self, extern_data_raw: Dict[str, torch.Tensor], *, train_flag: boo
...
         rf.init_train_step_run_ctx(train_flag=train_flag)
 
-        with autocast(device_type=self._device, dtype=self._autocast_dtype) if self._use_autocast else nullcontext():
-            sentinel_kw = {"__fwd_compatible_random_arg_%i" % int(random() * 100): None}
-            self._train_step_func(model=self._orig_model, extern_data=extern_data, **sentinel_kw)
+        sentinel_kw = {"__fwd_compatible_random_arg_%i" % int(random() * 100): None}
+
+        with autocast(
+            device_type=self._device, dtype=self._autocast_dtype
+        ) if self._use_autocast else nullcontext():
```
Comment (albertz) [inline returnn/torch/engine.py:445]:
```suggestion
        ) if self._use_autocast else nullcontext(), ... if self._use_torch_distributed else nullcontext():
```
Like this. (Properly formatted.)

File: returnn/torch/engine.py [hunk 16]
```diff
@@ -8,6 +8,10 @@
 
 import os
 import torch
+import time
+from torch.distributed import init_process_group, destroy_process_group
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.distributed.algorithms.join import Join
```
Comment (albertz) [inline returnn/torch/engine.py:14]: Do not import such generic names (`Join`) into the main namespace.

Comment (albertz) [reply returnn/torch/engine.py:14]: Do you even need this here? I thought this is now in `returnn.torch.distributed`?

File: returnn/torch/engine.py [hunk 17]
```diff
@@ -193,13 +250,29 @@ def train_epoch(self):
                 {name: float(_to_raw(loss.get_inv_norm_factor())) for name, loss in train_ctx.losses.items()}
             )
 
-            if self._grad_scaler is not None:
-                self._grad_scaler.scale(total_loss).backward()
-                self._grad_scaler.step(self._updater.get_optimizer())
-                self._grad_scaler.update()
+            if self._use_torch_distributed and (step_idx % self._accum_grad_multiple_step) != (
```
Comment (albertz) [inline returnn/torch/engine.py:253]: Do not put the `if` statement here. Do not repeat the same code below twice. Rather, Just have the code once, but have the context manager optionally, and use `nullcontext` otherwise.

File: returnn/torch/distributed.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+            return None
+    _is_set_up = True
+    if config.typed_value("torch_distributed") is None:
+        return None
+    _ctx = DistributedContext(config=config)
+    return _ctx
+
+
+@contextmanager
+def _ddp_ctx(pt_model):
```
Comment (albertz) [inline returnn/torch/distributed.py:83]: As you use this outside of this module, make it public. Also choose a more describing name, like `ddp_train_ctx` or so.

File: returnn/torch/engine.py [hunk 19]
```diff
@@ -105,6 +119,14 @@ def init_train_from_config(
         """
         assert config is self.config
         super().init_train_from_config(config=config)
+        if self._use_torch_distributed:
+            import returnn.torch.distributed
+
+            torch_distributed = returnn.torch.distributed.get_ctx(config=config)
+            local_rank = torch_distributed.local_rank()
+            print(f"Start running basic DDP example on local rank {local_rank}.", file=log.v2)
```
Comment (albertz) [inline returnn/torch/engine.py:127]: Why "DDP example"?

File: returnn/torch/engine.py [hunk 20]
```diff
@@ -130,14 +152,23 @@ def init_train_from_config(
         self._load_model(epoch=self._start_epoch)
         self._save_model_epoch_interval = config.int("save_interval", 1)
 
-        self._updater = Updater(self.config, self._pt_model, self.learning_rate)
+        if self._use_torch_distributed:
```
Comment (albertz) [inline returnn/torch/engine.py:155]: `self._ddp_pt_model = None` should be set outside/before this.

File: returnn/torch/engine.py [hunk 21]
```diff
@@ -155,6 +186,8 @@ def train(self):
             self.epoch += 1
             self._epoch_mp_shared.value = self.epoch
 
+        if self._use_torch_distributed:
+            destroy_process_group()
```
Comment (albertz) [inline returnn/torch/engine.py:190]: Why are you calling this here? As far as I understood, this function should be called at the end of the process? E.g. in our `finalize` function.

File: returnn/torch/engine.py [hunk 22]
```diff
@@ -130,14 +152,23 @@ def init_train_from_config(
         self._load_model(epoch=self._start_epoch)
         self._save_model_epoch_interval = config.int("save_interval", 1)
 
-        self._updater = Updater(self.config, self._pt_model, self.learning_rate)
+        if self._use_torch_distributed:
+            # wrap the model use torch distributed class
+            self._ddp_pt_model = self._torch_distributed_class(
+                self._pt_model, device_ids=[int(os.environ["LOCAL_RANK"])], **self._torch_distributed_options
```
[the remark this answers is not in the document]
Comment (albertz) [reply returnn/torch/engine.py:158]: After discussion:

The reason `os.environ` was used is because this was the only known way to get the local rank (as opposed to the global rank).

It's a bit ambiguous what we should use this, whether it is the local rank, global rank, the device id 0, or what else. It also matter how `CUDA_VISIBLE_DEVICES` is set up, single node vs multi node training, etc. Also how it is run later on, via SGE, mpirun, whatever. Using the local rank is probably reasonable for our current use cases though.

The next question is, how to get the local rank. `torch.distributed` does not provide such function?

So, we agreed to introduce some helper functions (in our `returnn.torch.distributed` module):
```python
def get_device_ids():
  # It depends on the specific setup what to return here,
  # how CUDA_VISIBLE_DEVICES is set up, etc.
  # This is currently a reasonable assumption
  # but we might extend the logic later,
  # or make it configurable.
  return [get_local_rank()]

def get_local_rank():
  # torch.distributed does not seem to provide a function for this.
  # Via mpirun (OpenMPI), this env variable would be set.
  # It should fail with an error otherwise.
  return int(os.environ["LOCAL_RANK"])
```

File: returnn/torch/engine.py [hunk 23]
```diff
@@ -177,11 +210,35 @@ def train_epoch(self):
...
-        for data in self._train_dataloader:
-            self._updater.get_optimizer().zero_grad()
+        epoch_start_time = time.time()
+
+        data_iter = iter(self._train_dataloader)
+        elapsed_computation_time = 0
+
+        while True:
+            data = next(data_iter, None)
+            _has_data = torch.tensor([data is not None], dtype=torch.int8).to(self._device)
```
[the remark this answers is not in the document]
Comment (Judyxujj) [reply returnn/torch/engine.py:220]: if I do not use `to(self._device)`, I get the error RuntimeError: Tensors must be CUDA and dense

[the remark this answers is not in the document]
Comment (albertz) [reply returnn/torch/engine.py:220]: Where do you get that error?

[the remark this answers is not in the document]
Comment (Judyxujj) [reply returnn/torch/engine.py:218]: from here `torch.distributed.all_reduce(_has_data, op=torch.distributed.ReduceOp.MIN)`

File: returnn/torch/engine.py [hunk 24]
```diff
@@ -130,14 +152,23 @@ def init_train_from_config(
...
 
-        self._updater = Updater(self.config, self._pt_model, self.learning_rate)
+        if self._use_torch_distributed:
+            # wrap the model use torch distributed class
+            self._ddp_pt_model = self._torch_distributed_class(
+                self._pt_model, device_ids=[int(os.environ["LOCAL_RANK"])], **self._torch_distributed_options
+            )
+            self._updater = Updater(self.config, self._ddp_pt_model, self.learning_rate)
+        else:
+            self._updater = Updater(self.config, self._pt_model, self.learning_rate)
```
[the remark this answers is not in the document]
Comment (Judyxujj) [reply returnn/torch/engine.py]: I have tested, to pass the DDP wrapped module or not has indeed no impact, so now I just pass `self._pt_mode ` here

File: returnn/torch/engine.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (Judyxujj) [reply returnn/torch/engine.py]: as suggested, I added the helper functions in `returnn.torch.distributed` module

File: returnn/torch/engine.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (albertz) [reply returnn/torch/engine.py:218]: Oh ok. I did not expect that it can only use one single mechanism for communication, namely NCCL here. This might be a bit suboptimal here (roundtrip CPU -> GPU -> NCCL -> GPU -> CPU; I was rather expecting CPU -> ethernet -> CPU). But some big warning comment then, and maybe we later need to deal with this.

[the remark this answers is not in the document]
Comment (albertz) [reply returnn/torch/engine.py:218]:
```suggestion
            # WARNING: torch.distributed works only for the registered device,
            # as it uses only one mechanism for communication, like NCCL.
            # This is suboptimal here as we have the roundtrip CPU -> GPU -> NCCL -> GPU -> CPU.
            # TODO: Use more direct CPU -> Ethernet -> CPU communication.
            _has_data = torch.tensor([data is not None], dtype=torch.int8).to(self._device)
```

[the remark this answers is not in the document]
Comment (Judyxujj) [reply returnn/torch/engine.py:222]: okay, I have added the warning comment here

File: returnn/torch/distributed.py [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+
+    def size(self):
+        """
+        :rtype: int
+        """
+        return self._size
+
+
+_is_set_up = False
+_ctx = None  # type: typing.Optional[HorovodContext]
```
Comment (albertz) [inline returnn/torch/distributed.py:56]:
```suggestion
_ctx = None  # type: Optional[DistributedContext]
```

**Comment by Judyxujj** [comment]
> There are conflicts. Can you rebase?
> 
> There are failing tests. Can you make sure they are passing?

I have merged the master branch to this branch and passed all tests now. In the commit [738b091](https://github.com/rwth-i6/returnn/commit/738b0910ee35d3d8ec3328a23a21eb9f3a249956) , `seq_tag` is added to the dataset key, it will cause error when doing the chunking. So I added an if statement to do chunking only on `data_dict['data']` as in commit [33cc459](https://github.com/rwth-i6/returnn/pull/1335/commits/33cc459dc4abfaba93bf2141fcb338e0737dd37e#). I am not sure if this is the right solution for it.
```
data = data_dict[data_key]
chunks = [
     data[start_index : start_index + chunk_size] for start_index in range(0, len(data), chunk_step)
 ]
```

File: returnn/torch/data/pipeline.py [hunk 26]
```diff
@@ -114,22 +114,23 @@ def __iter__(self):
...
-
-                if num_chunks is None:
-                    num_chunks = len(chunks)
-                else:
-                    assert num_chunks == len(
-                        chunks
-                    ), "Chunking resulted in different number of chunks for different data keys."
-
-                data_chunks[data_key] = chunks
+                if data_key == "data":
```
Comment (albertz) [inline returnn/torch/data/pipeline.py:117]: You never should hardcode any data key name here (except maybe of the special keys like `seq_tag`, but `data` is really not special). Also, there should be no logic specific for some data key (except maybe the special keys like `seq_tag`).

**Comment by albertz** [comment]
> I have merged the master branch to this branch and passed all tests now. In the commit [738b091](https://github.com/rwth-i6/returnn/commit/738b0910ee35d3d8ec3328a23a21eb9f3a249956) , `seq_tag` is added to the dataset key, it will cause error when doing the chunking. So I added an if statement to do chunking only on `data_dict['data']` as in commit [33cc459](https://github.com/rwth-i6/returnn/pull/1335/commits/33cc459dc4abfaba93bf2141fcb338e0737dd37e#). I am not sure if this is the right solution for it.

No it's not. I commented already.

But it's anyway not really related to this PR. This is a completely separate issue. Can you open a separate issue on it?

File: returnn/torch/data/pipeline.py (hunk 26 above)
Comment (albertz) [reply returnn/torch/data/pipeline.py:117]: I see you added this as a workaround for chunking, which breaks because of `seq_tag`. This does not belong here into the PR. Can you undo this change?

For testing now without having this fixed, just don't use chunking. Let us fix chunking in a separate commit or PR.

**Review by albertz — CHANGES_REQUESTED** [review]
Remove the data pipeline change. Otherwise I think we can merge now and then improve the code further when needed.

**Comment by Judyxujj** [comment]
I removed the data pipeline changes and opened a separate issue for that. All the tests are passed and this branch is ready to be merged.

**Review by albertz — APPROVED** [review]

**Comment by albertz** [comment]
Why do you actually have `_torch_distributed_class` (`class` in `torch_distributed` config dict) configurable? I recently changed the default of this to be `DistributedDataParallel`, but I wonder, are there really any potential cases where this can be sth different?

**Comment by Judyxujj** [comment]
What about the fully sharded data parallel (FSDP)?(https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html) 
But I think to support this, we need to modify more things
